from unittest import TestCase

from django.core.exceptions import ValidationError
from django.db import models


class ValidationMessagesTest(TestCase):

    def _test_validation_messages(self, field, value, expected):
        """
        Tests the validation messages for a given field and value.
        
        Args:
        field (Field): The field to test.
        value (Any): The value to be validated.
        expected (list): The expected list of validation error messages.
        
        Raises:
        ValidationError: If the validation messages do not match the expected ones.
        
        This function is used to ensure that the validation messages generated by a field match the expected messages when a specific value is passed to its clean method.
        """

        with self.assertRaises(ValidationError) as cm:
            field.clean(value, None)
        self.assertEqual(cm.exception.messages, expected)

    def test_autofield_field_raises_error_message(self):
        f = models.AutoField(primary_key=True)
        self._test_validation_messages(f, 'fõo', ['“fõo” value must be an integer.'])

    def test_integer_field_raises_error_message(self):
        f = models.IntegerField()
        self._test_validation_messages(f, 'fõo', ['“fõo” value must be an integer.'])

    def test_boolean_field_raises_error_message(self):
        f = models.BooleanField()
        self._test_validation_messages(f, 'fõo', ['“fõo” value must be either True or False.'])

    def test_nullable_boolean_field_raises_error_message(self):
        f = models.BooleanField(null=True)
        self._test_validation_messages(f, 'fõo', ['“fõo” value must be either True, False, or None.'])

    def test_float_field_raises_error_message(self):
        f = models.FloatField()
        self._test_validation_messages(f, 'fõo', ['“fõo” value must be a float.'])

    def test_decimal_field_raises_error_message(self):
        f = models.DecimalField()
        self._test_validation_messages(f, 'fõo', ['“fõo” value must be a decimal number.'])

    def test_null_boolean_field_raises_error_message(self):
        f = models.NullBooleanField()
        self._test_validation_messages(f, 'fõo', ['“fõo” value must be either None, True or False.'])

    def test_date_field_raises_error_message(self):
        f = models.DateField()
        self._test_validation_messages(
            f, 'fõo',
            ['“fõo” value has an invalid date format. It must be in YYYY-MM-DD format.']
        )
        self._test_validation_messages(
            f, 'aaaa-10-10',
            ['“aaaa-10-10” value has an invalid date format. It must be in YYYY-MM-DD format.']
        )
        self._test_validation_messages(
            f, '2011-13-10',
            ['“2011-13-10” value has the correct format (YYYY-MM-DD) but it is an invalid date.']
        )
        self._test_validation_messages(
            f, '2011-10-32',
            ['“2011-10-32” value has the correct format (YYYY-MM-DD) but it is an invalid date.']
        )

    def test_datetime_field_raises_error_message(self):
        """
        Test the validation of a DateTimeField in a Django model.
        
        This function tests the validation of a DateTimeField by providing it with various input values and checking if the expected error messages are raised.
        
        Parameters:
        f (DateTimeField): The DateTimeField instance to test.
        
        Test Cases:
        1. Test with a wrong format: 'fõo' should raise an error message indicating an invalid format.
        2. Test with a correct format but an invalid date: '2011-10-
        """

        f = models.DateTimeField()
        # Wrong format
        self._test_validation_messages(
            f, 'fõo',
            ['“fõo” value has an invalid format. It must be in YYYY-MM-DD HH:MM[:ss[.uuuuuu]][TZ] format.']
        )
        # Correct format but invalid date
        self._test_validation_messages(
            f, '2011-10-32',
            ['“2011-10-32” value has the correct format (YYYY-MM-DD) but it is an invalid date.']
        )
        # Correct format but invalid date/time
        self._test_validation_messages(
            f, '2011-10-32 10:10',
            ['“2011-10-32 10:10” value has the correct format (YYYY-MM-DD HH:MM[:ss[.uuuuuu]][TZ]) '
             'but it is an invalid date/time.']
        )

    def test_time_field_raises_error_message(self):
        f = models.TimeField()
        # Wrong format
        self._test_validation_messages(
            f, 'fõo',
            ['“fõo” value has an invalid format. It must be in HH:MM[:ss[.uuuuuu]] format.']
        )
        # Correct format but invalid time
        self._test_validation_messages(
            f, '25:50',
            ['“25:50” value has the correct format (HH:MM[:ss[.uuuuuu]]) but it is an invalid time.']
        )
